File: apps/android/app/src/main/java/com/signalasi/chat/AgentMemoryTrust.kt
package com.signalasi.chat

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest
import java.util.UUID

data class AgentMemoryUsageRecord(
    val id: String = UUID.randomUUID().toString(),
    val memoryIds: List<String>,
    val conversationId: String,
    val turnId: String,
    val querySha256: String,
    val runId: String = "",
    val answerPreview: String = "",
    val oldestMemoryTimestampMillis: Long = 0L,
    val newestMemoryTimestampMillis: Long = 0L,
    val selectedAtMillis: Long = System.currentTimeMillis(),
    val answeredAtMillis: Long = 0L
)

data class AgentMemoryTrustProfile(
    val memoryId: String,
    val whyRemembered: String,
    val source: String,
    val currentState: String,
    val lastVerifiedAtMillis: Long,
    val privateMemory: Boolean,
    val usages: List<AgentMemoryUsageRecord>
)

internal object AgentMemoryHorizonPolicy {
    fun qualifies(
        oldestMemoryTimestampMillis: Long,
        answeredAtMillis: Long,
        requiredHorizonDays: Int
    ): Boolean {
        val horizon = requiredHorizonDays.coerceIn(0, MAX_HORIZON_DAYS)
        if (horizon == 0) return true
        if (oldestMemoryTimestampMillis <= 0L || answeredAtMillis <= 0L) return false
        val minimumAgeMillis = horizon.toLong() * DAY_MILLIS
        return answeredAtMillis - oldestMemoryTimestampMillis >= minimumAgeMillis
    }

    private const val MAX_HORIZON_DAYS = 3_650
    private const val DAY_MILLIS = 86_400_000L
}

class AgentMemoryTrustStore internal constructor(private val database: AgentEncryptedDatabase) {
    constructor(context: Context) : this(AgentEncryptedDatabase(context.applicationContext, DATABASE))

    @Synchronized
    fun recordSelection(
        memoryIds: List<String>,
        conversationId: String,
        turnId: String,
        query: String,
        memoryTimestampsMillis: List<Long> = emptyList(),
        nowMillis: Long = System.currentTimeMillis()
    ): AgentMemoryUsageRecord? {
        val ids = memoryIds.map(String::trim).filter(String::isNotBlank).distinct().take(MAX_MEMORY_IDS)
        val cleanConversation = conversationId.trim().take(160)
        val cleanTurn = turnId.trim().take(160)
        if (ids.isEmpty() || cleanConversation.isBlank()) return null
        val digest = queryDigest(query)
        val pendingKey = pendingIndexKey(cleanConversation, cleanTurn, digest)
        val pending = indexedRecords(pendingKey)
        val duplicate = pending.firstOrNull { item ->
            item.conversationId == cleanConversation &&
                item.turnId == cleanTurn &&
                item.querySha256 == digest &&
                item.memoryIds == ids &&
                nowMillis - item.selectedAtMillis in 0..DUPLICATE_WINDOW_MILLIS
        }
        if (duplicate != null) return duplicate
        val record = AgentMemoryUsageRecord(
            memoryIds = ids,
            conversationId = cleanConversation,
            turnId = cleanTurn,
            querySha256 = digest,
            oldestMemoryTimestampMillis = memoryTimestampsMillis.filter { it > 0L }.minOrNull() ?: 0L,
            newestMemoryTimestampMillis = memoryTimestampsMillis.filter { it > 0L }.maxOrNull() ?: 0L,
            selectedAtMillis = nowMillis
        )
        database.mutateStrings(mapOf(
            "$KEY_PREFIX${record.id}" to encode(record).toString(),
            pendingKey to encodeIds((pending.map(AgentMemoryUsageRecord::id) + record.id).distinct())
        ))
        prune()
        return record
    }

    fun hasPendingSelection(conversationId: String, turnId: String, query: String): Boolean {
        val cleanConversation = conversationId.trim().take(160)
        if (cleanConversation.isBlank()) return false
        return database.contains(pendingIndexKey(
            cleanConversation,
            turnId.trim().take(160),
            queryDigest(query)
        ))
    }

    @Synchronized
    fun attachAnswer(
        conversationId: String,
        runId: String,
        answer: String,
        query: String = "",
        turnId: String = "",
        answeredAtMillis: Long = System.currentTimeMillis()
    ): Int {
        val cleanConversation = conversationId.trim()
        val cleanRun = runId.trim()
        if (cleanConversation.isBlank() || cleanRun.isBlank()) return 0
        val queryDigest = query.trim().takeIf(String::isNotBlank)?.let(::queryDigest)
        val cleanTurn = turnId.trim().take(160)
        val pendingKey = queryDigest?.let { pendingIndexKey(cleanConversation, cleanTurn, it) }
        val indexed = pendingKey?.let(::indexedRecords).orEmpty()
        val candidates = indexed.filter { record ->
            record.conversationId == cleanConversation &&
                record.runId.isBlank() &&
                (cleanTurn.isBlank() || record.turnId == cleanTurn) &&
                (queryDigest == null || record.querySha256 == queryDigest) &&
                answeredAtMillis - record.selectedAtMillis in 0..ANSWER_LINK_WINDOW_MILLIS
        }
        if (candidates.isEmpty()) return 0
        val linked = candidates.map { record ->
            record.copy(
                runId = cleanRun.take(160),
                answerPreview = answer.replace(Regex("\\s+"), " ").trim().take(MAX_ANSWER_PREVIEW),
                answeredAtMillis = answeredAtMillis
            )
        }
        val existingRunIds = indexedIds(runIndexKey(cleanRun))
        database.mutateStrings(
            upserts = buildMap {
                linked.forEach { record -> put("$KEY_PREFIX${record.id}", encode(record).toString()) }
                put(runIndexKey(cleanRun), encodeIds((existingRunIds + linked.map(AgentMemoryUsageRecord::id)).distinct()))
            },
            removeKeys = listOfNotNull(pendingKey)
        )
        return linked.size
    }

    @Synchronized
    fun usagesFor(memoryId: String, limit: Int = 20): List<AgentMemoryUsageRecord> {
        val cleanId = memoryId.trim()
        if (cleanId.isBlank()) return emptyList()
        return recent(MAX_RECORDS).filter { cleanId in it.memoryIds }.take(limit.coerceIn(1, 100))
    }

    @Synchronized
    fun verifiedUsageForRun(
        runId: String,
        requiredHorizonDays: Int,
        answeredAtMillis: Long
    ): AgentMemoryUsageRecord? {
        val cleanRunId = runId.trim()
        if (cleanRunId.isBlank()) return null
        return indexedRecords(runIndexKey(cleanRunId)).firstOrNull { record ->
            record.runId == cleanRunId &&
                record.memoryIds.isNotEmpty() &&
                record.answeredAtMillis > 0L &&
                AgentMemoryHorizonPolicy.qualifies(
                    record.oldestMemoryTimestampMillis,
                    answeredAtMillis,
                    requiredHorizonDays
                )
        }
    }

    @Synchronized
    fun recent(limit: Int = 100): List<AgentMemoryUsageRecord> {
        val keys = database.recentKeys(KEY_PREFIX, limit.coerceIn(1, MAX_RECORDS))
        val values = database.readStrings(keys)
        return keys.mapNotNull { key -> values[key]?.let(::decode) }
            .sortedByDescending(AgentMemoryUsageRecord::selectedAtMillis)
    }

    fun profile(item: AgentMemoryItem): AgentMemoryTrustProfile = AgentMemoryTrustProfile(
        memoryId = item.id,
        whyRemembered = item.whyRemembered.ifBlank { whyFromSource(item) },
        source = item.source,
        currentState = when (item.status) {
            AgentMemoryStatus.ACTIVE -> "current"
            AgentMemoryStatus.CONFLICTED -> "conflicted"
            AgentMemoryStatus.SUPERSEDED -> "historical"
        },
        lastVerifiedAtMillis = maxOf(item.lastConfirmedAtMillis, item.timestampMillis),
        privateMemory = item.privateMemory,
        usages = usagesFor(item.id)
    )

    private fun whyFromSource(item: AgentMemoryItem): String = when (item.source) {
        "explicit_core_memory" -> "The user explicitly stated this durable fact."
        "automatic_learning" -> "Repeated successful Agent runs supported this memory."
        "automatic_failure_learning" -> "Repeated failures established a condition that should not be retried unchanged."
        "memory_edit" -> "The user corrected an earlier memory."
        "memory_conflict_selection", "memory_conflict_merge" -> "The user resolved conflicting memory candidates."
        else -> if (item.autoLearned) "SignalASI inferred this from repeated evidence." else "Saved from an Agent interaction."
    }

    private fun prune() {
        val retained = database.recentKeys(KEY_PREFIX, MAX_RECORDS).toHashSet()
        database.removeAll(database.keys(KEY_PREFIX).filterNot(retained::contains))
    }

    private fun indexedRecords(indexKey: String): List<AgentMemoryUsageRecord> {
        val ids = indexedIds(indexKey)
        if (ids.isEmpty()) return emptyList()
        val keys = ids.map { "$KEY_PREFIX$it" }
        val values = database.readStrings(keys)
        return keys.mapNotNull { key -> values[key]?.let(::decode) }
    }

    private fun indexedIds(indexKey: String): List<String> = runCatching {
        val array = JSONArray(database.readString(indexKey, "[]"))
        buildList {
            for (index in 0 until array.length()) {
                array.optString(index).trim().takeIf(String::isNotBlank)?.let(::add)
            }
        }.distinct().take(MAX_INDEXED_RECORDS)
    }.getOrDefault(emptyList())

    private fun encodeIds(ids: List<String>) = JSONArray(ids.distinct().takeLast(MAX_INDEXED_RECORDS)).toString()

    private fun pendingIndexKey(conversationId: String, turnId: String, queryDigest: String): String =
        "$PENDING_INDEX_PREFIX${sha256("$conversationId\u001f${turnId.trim()}\u001f$queryDigest")}"

    private fun runIndexKey(runId: String) = "$RUN_INDEX_PREFIX${runId.trim()}"

    private fun queryDigest(query: String) = sha256(query.trim().take(MAX_QUERY_CHARS))

    private fun encode(value: AgentMemoryUsageRecord) = JSONObject()
        .put("id", value.id)
        .put("memory_ids", JSONArray(value.memoryIds))
        .put("conversation_id", value.conversationId)
        .put("turn_id", value.turnId)
        .put("query_sha256", value.querySha256)
        .put("run_id", value.runId)
        .put("answer_preview", value.answerPreview)
        .put("oldest_memory_timestamp_millis", value.oldestMemoryTimestampMillis)
        .put("newest_memory_timestamp_millis", value.newestMemoryTimestampMillis)
        .put("selected_at_millis", value.selectedAtMillis)
        .put("answered_at_millis", value.answeredAtMillis)

    private fun decode(raw: String): AgentMemoryUsageRecord? = runCatching {
        val json = JSONObject(raw)
        val ids = json.getJSONArray("memory_ids")
        AgentMemoryUsageRecord(
            id = json.getString("id"),
            memoryIds = buildList {
                for (index in 0 until ids.length()) ids.optString(index).takeIf(String::isNotBlank)?.let(::add)
            },
            conversationId = json.getString("conversation_id"),
            turnId = json.optString("turn_id"),
            querySha256 = json.getString("query_sha256"),
            runId = json.optString("run_id"),
            answerPreview = json.optString("answer_preview"),
            oldestMemoryTimestampMillis = json.optLong("oldest_memory_timestamp_millis"),
            newestMemoryTimestampMillis = json.optLong("newest_memory_timestamp_millis"),
            selectedAtMillis = json.optLong("selected_at_millis"),
            answeredAtMillis = json.optLong("answered_at_millis")
        )
    }.getOrNull()

    private fun sha256(value: String): String = MessageDigest.getInstance("SHA-256")
        .digest(value.toByteArray(Charsets.UTF_8))
        .joinToString("") { "%02x".format(it) }

    private companion object {
        const val DATABASE = "signalasi_memory_trust_v1"
        const val KEY_PREFIX = "usage:"
        const val PENDING_INDEX_PREFIX = "pending-index:"
        const val RUN_INDEX_PREFIX = "run-index:"
        const val MAX_RECORDS = 2_000
        const val MAX_INDEXED_RECORDS = 32
        const val MAX_QUERY_CHARS = 8_000
        const val MAX_MEMORY_IDS = 32
        const val MAX_ANSWER_PREVIEW = 320
        const val DUPLICATE_WINDOW_MILLIS = 60_000L
        const val ANSWER_LINK_WINDOW_MILLIS = 30L * 60_000L
    }
}

File: apps/ios/GalaxySSI/AgentMemoryTrust.swift
import CryptoKit
import Foundation

struct AgentMemoryUsageRecord: Codable, Equatable, Identifiable {
  var id: String
  var memoryIds: [String]
  var conversationId: String
  var turnId: String
  var querySHA256: String
  var runId: String
  var answerPreview: String
  var selectedAtMillis: Int64
  var answeredAtMillis: Int64

  enum CodingKeys: String, CodingKey {
    case id
    case memoryIds = "memory_ids"
    case conversationId = "conversation_id"
    case turnId = "turn_id"
    case querySHA256 = "query_sha256"
    case runId = "run_id"
    case answerPreview = "answer_preview"
    case selectedAtMillis = "selected_at_millis"
    case answeredAtMillis = "answered_at_millis"
  }

  init(
    id: String = UUID().uuidString,
    memoryIds: [String],
    conversationId: String,
    turnId: String = "",
    querySHA256: String,
    runId: String = "",
    answerPreview: String = "",
    selectedAtMillis: Int64,
    answeredAtMillis: Int64 = 0
  ) {
    self.id = id
    self.memoryIds = Array(Set(memoryIds.map(Self.clean).filter { !$0.isEmpty })).sorted()
    self.conversationId = Self.clean(conversationId)
    self.turnId = Self.clean(turnId)
    self.querySHA256 = Self.clean(querySHA256)
    self.runId = Self.clean(runId)
    self.answerPreview = String(answerPreview.trimmingCharacters(in: .whitespacesAndNewlines).prefix(320))
    self.selectedAtMillis = max(selectedAtMillis, 0)
    self.answeredAtMillis = max(answeredAtMillis, 0)
  }

  private static func clean(_ value: String) -> String {
    String(value.trimmingCharacters(in: .whitespacesAndNewlines).prefix(256))
  }
}

struct AgentMemoryTrustProfile: Codable, Equatable, Identifiable {
  var id: String { memoryId }
  var memoryId: String
  var whyRemembered: String
  var source: String
  var currentState: String
  var lastVerifiedAtMillis: Int64
  var privateMemory: Bool
  var usages: [AgentMemoryUsageRecord]

  enum CodingKeys: String, CodingKey {
    case memoryId = "memory_id"
    case whyRemembered = "why_remembered"
    case source
    case currentState = "current_state"
    case lastVerifiedAtMillis = "last_verified_at_millis"
    case privateMemory = "private_memory"
    case usages
  }
}

final class AgentMemoryTrustStore {
  private struct State: Codable {
    var usages: [AgentMemoryUsageRecord] = []
  }

  static let defaultKey = "galaxyssi-ios-agent-memory-trust-v1"
  private let defaults: UserDefaults
  private let secrets: GalaxySSISecretStore
  private let key: String
  private let nowMillis: () -> Int64
  private let lock = NSRecursiveLock()

  init(
    defaults: UserDefaults = .standard,
    secrets: GalaxySSISecretStore = KeychainSecretStore.shared,
    key: String = AgentMemoryTrustStore.defaultKey,
    nowMillis: @escaping () -> Int64 = AgentMemoryClock.nowMillis
  ) {
    self.defaults = defaults
    self.secrets = secrets
    self.key = key
    self.nowMillis = nowMillis
  }

  @discardableResult
  func recordSelection(
    memories: [AgentMemoryItem],
    conversationId: String,
    turnId: String = "",
    query: String,
    runId: String = ""
  ) -> AgentMemoryUsageRecord? {
    let memoryIds = Array(memories.filter { !$0.privateMemory }.map(\.id).prefix(32))
    let conversationId = String(conversationId.trimmingCharacters(in: .whitespacesAndNewlines).prefix(160))
    let turnId = String(turnId.trimmingCharacters(in: .whitespacesAndNewlines).prefix(160))
    guard !memoryIds.isEmpty, !conversationId.isEmpty else { return nil }
    let selectedAt = nowMillis()
    let digest = Self.digest(String(query.trimmingCharacters(in: .whitespacesAndNewlines).prefix(8_000)))
    return locked {
      var state = load()
      if let existing = state.usages.last(where: {
        $0.conversationId == conversationId &&
          $0.turnId == turnId &&
          $0.querySHA256 == digest &&
          $0.memoryIds == Array(Set(memoryIds)).sorted() &&
          selectedAt - $0.selectedAtMillis <= Self.selectionDedupeMillis
      }) {
        return existing
      }
      let usage = AgentMemoryUsageRecord(
        memoryIds: memoryIds,
        conversationId: conversationId,
        turnId: turnId,
        querySHA256: digest,
        runId: runId,
        selectedAtMillis: selectedAt
      )
      state.usages.append(usage)
      state.usages = Array(state.usages.suffix(Self.maximumUsageRecords))
      save(state)
      return usage
    }
  }

  @discardableResult
  func attachAnswer(
    conversationId: String,
    runId: String = "",
    answer: String,
    answeredAtMillis: Int64? = nil
  ) -> Int {
    let conversationId = String(conversationId.trimmingCharacters(in: .whitespacesAndNewlines).prefix(160))
    let runId = String(runId.trimmingCharacters(in: .whitespacesAndNewlines).prefix(160))
    let answeredAt = answeredAtMillis ?? nowMillis()
    guard !conversationId.isEmpty, !runId.isEmpty else { return 0 }
    return locked {
      var state = load()
      let indices = state.usages.indices.filter { index in
        let usage = state.usages[index]
        return usage.conversationId == conversationId && (usage.runId.isEmpty || usage.runId == runId) &&
          answeredAt - usage.selectedAtMillis >= 0 &&
          answeredAt - usage.selectedAtMillis <= Self.answerWindowMillis
      }
      guard !indices.isEmpty else { return 0 }
      for index in indices {
        state.usages[index].answerPreview = String(answer
          .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
          .trimmingCharacters(in: .whitespacesAndNewlines)
          .prefix(Self.maximumAnswerPreviewCharacters))
        state.usages[index].answeredAtMillis = max(answeredAt, state.usages[index].selectedAtMillis)
        if state.usages[index].runId.isEmpty { state.usages[index].runId = runId }
      }
      save(state)
      return indices.count
    }
  }

  func usages(for memoryId: String, limit: Int = 50) -> [AgentMemoryUsageRecord] {
    let memoryId = memoryId.trimmingCharacters(in: .whitespacesAndNewlines)
    return locked {
      Array(load().usages.filter { $0.memoryIds.contains(memoryId) }
        .sorted { $0.selectedAtMillis > $1.selectedAtMillis }
        .prefix(min(max(limit, 1), 200)))
    }
  }

  func recent(limit: Int = 100) -> [AgentMemoryUsageRecord] {
    locked {
      Array(load().usages.sorted { $0.selectedAtMillis > $1.selectedAtMillis }
        .prefix(min(max(limit, 1), 500)))
    }
  }

  func profile(memory: AgentMemoryItem) -> AgentMemoryTrustProfile {
    let records = usages(for: memory.id)
    return AgentMemoryTrustProfile(
      memoryId: memory.id,
      whyRemembered: memory.whyRemembered.isEmpty ? Self.whyFromSource(memory.source) : memory.whyRemembered,
      source: memory.source,
      currentState: Self.currentState(memory.status),
      lastVerifiedAtMillis: max(memory.lastConfirmedAtMillis, memory.timestampMillis),
      privateMemory: memory.privateMemory,
      usages: records
    )
  }

  func clear() {
    locked {
      GalaxySSIEncryptedUserDefaultsStore.destroy(defaults: defaults, key: key, secrets: secrets)
    }
  }

  static func destroyPersistentStore(
    defaults: UserDefaults = .standard,
    secrets: GalaxySSISecretStore = KeychainSecretStore.shared,
    key: String = AgentMemoryTrustStore.defaultKey
  ) {
    GalaxySSIEncryptedUserDefaultsStore.destroy(defaults: defaults, key: key, secrets: secrets)
  }

  static func whyFromSource(_ source: String) -> String {
    switch source.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
    case "explicit_core_memory": return "The user explicitly stated this durable fact."
    case "automatic_learning": return "Repeated successful Agent runs supported this memory."
    case "automatic_failure_learning": return "Repeated failures established a condition that should not be retried unchanged."
    case "memory_edit": return "The user corrected an earlier memory."
    case "memory_conflict_selection", "memory_conflict_merge": return "The user resolved conflicting memory candidates."
    default: return "Saved from an Agent interaction."
    }
  }

  private static func currentState(_ status: AgentMemoryStatus) -> String {
    switch status {
    case .active: return "current"
    case .conflicted: return "conflicted"
    case .superseded: return "historical"
    }
  }

  private func load() -> State {
    guard let data = GalaxySSIEncryptedUserDefaultsStore.load(defaults: defaults, key: key, secrets: secrets),
          let state = try? JSONDecoder().decode(State.self, from: data) else { return State() }
    return state
  }

  private func save(_ state: State) {
    guard let data = try? JSONEncoder().encode(state) else { return }
    _ = GalaxySSIEncryptedUserDefaultsStore.write(data, defaults: defaults, key: key, secrets: secrets)
  }

  private func locked<T>(_ work: () -> T) -> T {
    lock.lock()
    defer { lock.unlock() }
    return work()
  }

  private static func digest(_ value: String) -> String {
    SHA256.hash(data: Data(value.utf8)).map { String(format: "%02x", $0) }.joined()
  }

  private static let selectionDedupeMillis: Int64 = 60_000
  private static let answerWindowMillis: Int64 = 30 * 60_000
  private static let maximumAnswerPreviewCharacters = 320
  private static let maximumUsageRecords = 2_000
}
